### ISSUE TITLE
"Error reading from database" when creating newsletter activity

### ISSUE THREAD

**Issue by jojoob** [dasistwas/moodle-mod_newsletter#25]
When I want to create a new newsletter activity in a course I get the following error:

```
Debug info: ERROR: invalid input syntax for integer: ""
SELECT f.id AS id, f.contenthash, f.pathnamehash, f.contextid, f.component, f.filearea, f.itemid, f.filepath, f.filename, f.userid, f.filesize, f.mimetype, f.status, f.source, f.author, f.license, f.timecreated, f.timemodified, f.sortorder, f.referencefileid, r.repositoryid AS repositoryid, r.reference AS reference, r.lastsync AS referencelastsync
FROM mdl_files f
LEFT JOIN mdl_files_reference r
ON f.referencefileid = r.id
WHERE f.contextid = $1
AND f.component = $2
AND f.filearea = $3
AND f.itemid = $4 ORDER BY itemid, filepath, filename
[array (
0 => 19,
1 => 'mod_newsletter',
2 => 'stylesheet',
3 => '',
)]
Error code: dmlreadexception
```

```
Stack trace:
line 474 of /lib/dml/moodle_database.php: dml_read_exception thrown
line 244 of /lib/dml/pgsql_native_moodle_database.php: call to moodle_database->query_end()
line 794 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->query_end()
line 827 of /lib/filestorage/file_storage.php: call to pgsql_native_moodle_database->get_records_sql()
line 397 of /lib/filelib.php: call to file_storage->get_area_files()
line 90 of /mod/newsletter/mod_form.php: call to file_prepare_draft_area()
line 204 of /lib/formslib.php: call to mod_newsletter_mod_form->definition()
line 95 of /course/moodleform_mod.php: call to moodleform->__construct()
line 255 of /course/modedit.php: call to moodleform_mod->__construct()
```

## Software info:
* Moodle version: 3.1.3+ (Build: 20161115)
* Database: Postgresql 9.3+

**Comment by dasistwas** [comment]
Thank you for reporting that. We will fix that issue.

**Comment by jojoob** [comment]
The problem is caused by using `$this->current->id` without an empty() check in [mod_form.php#L88](https://github.com/dasistwas/moodle-mod_newsletter/blob/9f9ef4790fe73b03e2a20377c1422d064613eda9/mod_form.php#L88). Using `empty($this->current->id) ? null : $this->current->id` instead of just `$this->current->id` works for me.

Side note: `'stylesheets'` is used as the element name at the [addElement statement](https://github.com/dasistwas/moodle-mod_newsletter/blob/9f9ef4790fe73b03e2a20377c1422d064613eda9/mod_form.php#L68) for the filemanager while the constant`NEWSLETTER_FILE_AREA_STYLESHEET` (which resolves to 'stylesheet' - without a s at the end) is used at the [`file_get_submitted_draft_itemid` and `file_prepare_draft_area` function calls](https://github.com/dasistwas/moodle-mod_newsletter/blob/9f9ef4790fe73b03e2a20377c1422d064613eda9/mod_form.php#L87-L89).

But this seems not to be the only problem with Postgresql compatibility: When an existing activity instance is viewed another SQL error occurs:

```
Error reading from database

More information about this error

Debug info: ERROR: argument of AND must be type boolean, not type integer
LINE 4: AND 1 AND 1 ORDER BY i.publishon DESC
^
SELECT i.*
FROM mdl_newsletter_issues i
WHERE i.newsletterid = $1
AND 1 AND 1 ORDER BY i.publishon DESC
[array (
0 => '1',
)]
Error code: dmlreadexception

Stack trace:
line 474 of /lib/dml/moodle_database.php: dml_read_exception thrown
line 244 of /lib/dml/pgsql_native_moodle_database.php: call to moodle_database->query_end()
line 794 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->query_end()
line 1234 of /mod/newsletter/locallib.php: call to pgsql_native_moodle_database->get_records_sql()
line 949 of /mod/newsletter/locallib.php: call to mod_newsletter->get_issues()
line 485 of /mod/newsletter/locallib.php: call to mod_newsletter->prepare_issue_list()
line 274 of /mod/newsletter/locallib.php: call to mod_newsletter->view_newsletter()
line 67 of /mod/newsletter/view.php: call to mod_newsletter->view()
```

**Comment by dasistwas** [comment]
Fixed in 613ab01276f919548fc229be2cd8dcdb7f5992d3

Cheeers,
David
